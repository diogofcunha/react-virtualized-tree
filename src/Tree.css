.tree-node {
  user-select: none;
  cursor: pointer;
}
