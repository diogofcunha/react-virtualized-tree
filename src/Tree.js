import React from 'react';
import PropTypes from 'prop-types';
import { AutoSizer, List, CellMeasurerCache, CellMeasurer } from 'react-virtualized';

import { FlattenedNode } from './shapes/nodeShapes';

export default class Tree extends React.Component {
  _cache = new CellMeasurerCache({
    fixedWidth: true,
    minHeight: 20,
  })

  rowRenderer = ({ node, key, measure, style, NodeRenderer }) => {
    const {nodeMarginLeft} = this.props;

    return (
      <div
        key={key}
        className="tree-node"
        style={{ ...style, marginLeft: node.deepness * nodeMarginLeft}}
      >
        <NodeRenderer node={node} onChange={this.props.onChange} measure={measure}/>
      </div>
    );
  }

  measureRowRenderer =  (nodes) => ({ key, index, style, parent }) => {
    const { NodeRenderer } = this.props;
    const node = nodes[index];

    return (
      <CellMeasurer
        cache={this._cache}
        columnIndex={0}
        key={key}
        rowIndex={index}
        parent={parent}>
        {
          m => this.rowRenderer({ ...m, node, key, style, NodeRenderer })
        }
      </CellMeasurer>
    )
  }

  render() {
    const { nodes, rowHeight } = this.props;

    return (
      <AutoSizer>
        {({ height, width }) => (
          <List
            deferredMeasurementCache={this._cache}
            ref={r => this._list = r}
            height={height}
            rowCount={nodes.length}
            rowHeight={rowHeight || this._cache.rowHeight}
            rowRenderer={this.measureRowRenderer(nodes)}
            width={width}
          />
        )}
      </AutoSizer>
    );
  }
}

Tree.propTypes = {
  nodes: PropTypes.arrayOf(PropTypes.shape(FlattenedNode)).isRequired,
  NodeRenderer: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  nodeMarginLeft: PropTypes.number,
  rowHeight: PropTypes.number,
};
