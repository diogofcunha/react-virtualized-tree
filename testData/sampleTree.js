export const COLLAPSED_CHILDREN = {
  id: 3,
  name: 'Leaf 3',
  state: {
    expanded: false,
    favorite: true,
    deletable: true,
  },
  children: [
    {
      id: 'c-3',
      name: 'Leaf 3 Child',
      state: {},
    },
  ],
};

export const EXPANDED_CHILDREN = {
  id: 2,
  name: 'Leaf 2',
  state: {
    expanded: true,
    deletable: true,
  },
  children: [
    COLLAPSED_CHILDREN,
    {
      id: 4,
      name: 'Leaf 4',
    },
  ],
};

export const EXPANDED_NODE_IN_ROOT = {
  id: 0,
  name: 'Leaf 1',
  state: {
    expanded: true,
  },
  children: [
    EXPANDED_CHILDREN,
    {
      id: 5,
      name: 'Leaf 5',
    },
  ],
};

export const COLLAPSED_NODE_IN_ROOT = {
  id: 1,
  name: 'Leaf 6',
  state: {
    expanded: false,
    deletable: true,
  },
  children: [
    {
      id: 6,
      name: 'Leaf 7',
      state: {
        expanded: false,
      },
      children: [
        {
          id: 7,
          name: 'Leaf 8',
        },
        {
          id: 8,
          name: 'Leaf 9',
        },
      ],
    },
    {
      id: 9,
      name: 'Leaf 10',
    },
  ],
};

export const DELETABLE_IN_ROOT = {
  id: 'z',
  name: 'Leaf z',
  state: {
    deletable: true,
    favorite: true,
  },
};

export const DELETABLE_CHILDREN = EXPANDED_CHILDREN;

export const Nodes = [
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
  EXPANDED_NODE_IN_ROOT,
  COLLAPSED_NODE_IN_ROOT,
  DELETABLE_IN_ROOT,
];
