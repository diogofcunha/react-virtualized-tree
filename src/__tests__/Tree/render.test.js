jest.mock('react-virtualized');

import React from 'react';
import {render, cleanup, fireEvent} from 'react-testing-library';

import Tree from '../..';
import {Nodes} from '../../../testData/sampleTree';

class Example extends React.Component {
  state = {
    nodes: Nodes,
  };

  expandAllNodes = nodes => {
    return nodes.map(n => {
      let children;

      if (n.children) {
        children = this.expandAllNodes(n.children);
      }

      const state = n.state || {};

      return {
        ...n,
        children,
        state: {
          ...state,
          expanded: true,
        },
      };
    });
  };

  removePairRootPositionedNodes = () => {
    const nodes = this.state.nodes.filter((n, i) => i % 2 !== 0);

    this.setState({nodes});
  };

  render() {
    return (
      <div>
        <button
          data-testid="expand-all"
          onClick={() =>
            this.setState({
              nodes: this.expandAllNodes(this.state.nodes),
            })
          }
        />
        <button data-testid="remove-in-pair-root-pos" onClick={this.removePairRootPositionedNodes} />
        <div data-testid="tree">
          <Tree nodes={this.state.nodes}>
            {({style, node, ...rest}) => (
              <div style={style}>
                <span data-testid={`${node.id}`} data-haschildren={Boolean(node.children && node.children.length)}>
                  {node.name}
                </span>
              </div>
            )}
          </Tree>
        </div>
      </div>
    );
  }
}

describe('Tree rendering', () => {
  afterEach(cleanup);

  test('should render as expected on mount', () => {
    const {getByTestId} = render(<Example />);

    expect(getByTestId('tree')).toMatchSnapshot();
  });

  test('should render as expected on updates', () => {
    const {getByTestId} = render(<Example />);

    fireEvent.click(getByTestId('expand-all'));

    expect(getByTestId('tree')).toMatchSnapshot('all-expanded');

    fireEvent.click(getByTestId('remove-in-pair-root-pos'));

    expect(getByTestId('tree')).toMatchSnapshot('remove-in-pair-root-pos');
  });
});
