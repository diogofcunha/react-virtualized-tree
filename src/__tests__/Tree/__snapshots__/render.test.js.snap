// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree rendering should render as expected on mount 1`] = `
<div
  data-testid="tree"
>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="0"
    >
      Leaf 1
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="2"
    >
      Leaf 2
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="3"
    >
      Leaf 3
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="4"
    >
      Leaf 4
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="5"
    >
      Leaf 5
    </span>
  </div>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="1"
    >
      Leaf 6
    </span>
  </div>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="z"
    >
      Leaf z
    </span>
  </div>
</div>
`;

exports[`Tree rendering should render as expected on updates: all-expanded 1`] = `
<div
  data-testid="tree"
>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="0"
    >
      Leaf 1
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="2"
    >
      Leaf 2
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="3"
    >
      Leaf 3
    </span>
  </div>
  <div
    style="margin-left: 90px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="c-3"
    >
      Leaf 3 Child
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="4"
    >
      Leaf 4
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="5"
    >
      Leaf 5
    </span>
  </div>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="1"
    >
      Leaf 6
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="6"
    >
      Leaf 7
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="7"
    >
      Leaf 8
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="8"
    >
      Leaf 9
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="9"
    >
      Leaf 10
    </span>
  </div>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="z"
    >
      Leaf z
    </span>
  </div>
</div>
`;

exports[`Tree rendering should render as expected on updates: remove-in-pair-root-pos 1`] = `
<div
  data-testid="tree"
>
  <div
    style="margin-left: 0px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="1"
    >
      Leaf 6
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="true"
      data-testid="6"
    >
      Leaf 7
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="7"
    >
      Leaf 8
    </span>
  </div>
  <div
    style="margin-left: 60px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="8"
    >
      Leaf 9
    </span>
  </div>
  <div
    style="margin-left: 30px; user-select: none; cursor: pointer;"
  >
    <span
      data-haschildren="false"
      data-testid="9"
    >
      Leaf 10
    </span>
  </div>
</div>
`;
